File: lib/ex_scylla/statement/query.ex
defmodule ExScylla.Statement.Query do
  alias ExScylla.Types, as: T
  use ExScylla.Macros.Native, [
                          prefix: :q,
                          docs_rs_path: "/scylla/statement/query/struct.Query.html"
                        ]

  native_f func: :get_execution_profile_handle,
            args: [q],
            args_spec: [T.query()],
            return_spec: T.execution_profile_handle() | nil,
            doc_example: """
            iex> eph = ExecutionProfile.builder()
            ...>          |> ExecutionProfileBuilder.build()
            ...>          |> ExecutionProfile.into_handle()
            iex> q = Query.new("SELECT * FROM test;")
            ...>   |> Query.set_execution_profile_handle(eph)
            iex> q |> Query.get_execution_profile_handle()
            ...>   |> is_reference()
            true
            """

  native_f func: :set_execution_profile_handle,
            args: [q, profile_handle],
            args_spec: [T.query(), T.execution_profile_handle() | nil],
            return_spec: T.query(),
            doc_example: """
            iex> eph = ExecutionProfile.builder()
            ...>          |> ExecutionProfileBuilder.build()
            ...>          |> ExecutionProfile.into_handle()
            iex> Query.new("SELECT * FROM test;")
            ...>   |> Query.set_execution_profile_handle(eph)
            ...>   |> is_reference()
            true
            """

  native_f func: :get_request_timeout,
            args: [q],
            args_spec: [T.query()],
            return_spec: T.duration_ms() | nil,
            doc_example: """
            iex> q = Query.new("SELECT * FROM test;")
            ...>   |> Query.set_request_timeout(15000)
            iex> q |> Query.get_request_timeout()
            15000
            """

  native_f func: :set_request_timeout,
            args: [q, timeout_ms],
            args_spec: [T.query(), T.duration_ms() | nil],
            return_spec: T.query(),
            doc_example: """
            iex> q = Query.new("SELECT * FROM test;")
            ...>   |> Query.set_request_timeout(15000)
            iex> q |> Query.get_request_timeout()
            15000
            """

  native_f func: :disable_paging,
           args: [q],
           args_spec: [T.query()],
           return_spec: T.query(),
           doc_example: """
           iex> Query.new("SELECT * FROM test;")
           ...>   |> Query.disable_paging()
           ...>   |> is_reference()
           true
           """

  native_f func: :get_consistency,
           args: [q],
           args_spec: [T.query()],
           return_spec: T.consitency() | nil,
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> Query.get_consistency(q)
           nil
           iex> q = Query.set_consistency(q, :quorum)
           iex> Query.get_consistency(q)
           :quorum
           """

  native_f func: :get_is_idempotent,
           args: [q],
           args_spec: [T.query()],
           return_spec: boolean(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> false = Query.get_is_idempotent(q)
           iex> q = Query.set_is_idempotent(q, true)
           iex> true = Query.get_is_idempotent(q)
           """

  native_f func: :get_page_size,
           args: [q],
           args_spec: [T.query()],
           return_spec: pos_integer() | nil,
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> nil = Query.get_page_size(q)
           iex> q = Query.set_page_size(q, 10)
           iex> 10 = Query.get_page_size(q)
           """

  # TODO
  # native_f func: :get_retry_policy,
  #          args: [q],
  #          args_spec: [T.query()],
  #         return_spec: pos_integer() | nil

  native_f func: :get_serial_consistency,
           args: [q],
           args_spec: [T.query()],
           return_spec: T.serial_consistency() | nil,
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> nil = Query.get_serial_consistency(q)
           iex> q = Query.set_serial_consistency(q, :serial)
           iex> :serial = Query.get_serial_consistency(q)
           """

  native_f func: :get_timestamp,
           args: [q],
           args_spec: [T.query()],
           return_spec: T.ts_micros() | nil,
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> nil = Query.get_timestamp(q)
           iex> ts_micro = :os.system_time(:microsecond)
           iex> q = Query.set_timestamp(q, ts_micro)
           iex> ^ts_micro = Query.get_timestamp(q)
           """

  native_f func: :get_tracing,
           args: [q],
           args_spec: [T.query()],
           return_spec: boolean(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> false = Query.get_tracing(q)
           iex> q = Query.set_tracing(q, true)
           iex> true = Query.get_tracing(q)
           """

  native_f func: :new,
           args: [query_text],
           args_spec: [String.t()],
           return_spec: T.query(),
           doc_example: """
           iex> Query.new("SELECT * FROM test;")
           ...>   |> is_reference()
           true
           """

  native_f func: :set_consistency,
           args: [q, consistency],
           args_spec: [T.query(), T.consistency()],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_consistency(q, :one)
           iex> true = is_reference(q)
           """

  native_f func: :set_is_idempotent,
           args: [q, is_idempotent],
           args_spec: [T.query(), boolean()],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_is_idempotent(q, true)
           iex> true = is_reference(q)
           """

  native_f func: :set_page_size,
           args: [q, page_size],
           args_spec: [T.query(), pos_integer()],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_page_size(q, 10)
           iex> true = is_reference(q)
           """

  # TODO
  #native_f prefix: :q,
  #         func: :set_retry_policy,
  #         args: [q, page_size],
  #         args_spec: [T.query(), pos_integer()],
  #         return_spec: T.query()

  native_f func: :set_serial_consistency,
           args: [q, sc],
           args_spec: [T.query(), T.serial_consistency() | nil],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_serial_consistency(q, :local_serial)
           iex> true = is_reference(q)
           """

  native_f func: :set_timestamp,
           args: [q, timestamp_micros],
           args_spec: [T.query(), T.ts_micros() | nil],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_timestamp(q, :os.system_time(:microsecond))
           iex> true = is_reference(q)
           """

  native_f func: :set_tracing,
           args: [q, should_trace],
           args_spec: [T.query(), boolean()],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.set_tracing(q, true)
           iex> true = is_reference(q)
           """

  native_f func: :with_page_size,
           args: [q, page_size],
           args_spec: [T.query(), pos_integer()],
           return_spec: T.query(),
           doc_example: """
           iex> q = Query.new("SELECT * FROM test;")
           iex> q = Query.with_page_size(q, 10)
           iex> true = is_reference(q)
           """
end

File: native/ex_scylla/src/query.rs
pub mod types;
use std::time::Duration;

use crate::consts::*;
use crate::execution::execution_profile_handle::ExecutionProfileHandleResource;
use crate::types::*;
use crate::utils::*;
use rustler::{Atom, ResourceArc};
use scylla::query::Query;
use types::*;

#[rustler::nif]
fn q_get_execution_profile_handle(qr: ResourceArc<QueryResource>) -> Option<ResourceArc<ExecutionProfileHandleResource>> {
    let q: &Query = &qr.0;
    q.get_execution_profile_handle().map(|h| ResourceArc::new(ExecutionProfileHandleResource(h.clone())))
}

#[rustler::nif]
fn q_get_request_timeout(qr: ResourceArc<QueryResource>) -> Option<u64> {
    let q: &Query = &qr.0;
    q.get_request_timeout().map(|d| d.as_millis() as u64)
}

#[rustler::nif]
fn q_set_execution_profile_handle(q: ResourceArc<QueryResource>, profile_handle: Option<ResourceArc<ExecutionProfileHandleResource>>) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_execution_profile_handle(profile_handle.map(|ephr| ephr.0.clone()));
    q.ex()
}

#[rustler::nif]
fn q_set_request_timeout(q: ResourceArc<QueryResource>, timeout_ms: Option<u64>) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_request_timeout(timeout_ms.map(|ms| Duration::from_millis(ms)));
    q.ex()
}

#[rustler::nif]
fn q_disable_paging(q: ResourceArc<QueryResource>) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.disable_paging();
    q.ex()
}

#[rustler::nif]
fn q_get_consistency(q: ResourceArc<QueryResource>) -> Option<ScyllaConsistency> {
    let q: &Query = &q.0;
    q.get_consistency().map(|c| c.into())
}

#[rustler::nif]
fn q_get_is_idempotent(q: ResourceArc<QueryResource>) -> bool {
    let q: &Query = &q.0;
    q.get_is_idempotent()
}

#[rustler::nif]
fn q_get_page_size(q: ResourceArc<QueryResource>) -> Option<i32> {
    let q: &Query = &q.0;
    q.get_page_size()
}

#[rustler::nif]
fn q_get_retry_policy(_q: ResourceArc<QueryResource>) -> Atom {
    not_implemented_yet()
}

#[rustler::nif]
fn q_get_serial_consistency(q: ResourceArc<QueryResource>) -> Option<ScyllaSerialConsistency> {
    let q: &Query = &q.0;
    q.get_serial_consistency().map(|c| c.into())
}

#[rustler::nif]
fn q_get_timestamp(q: ResourceArc<QueryResource>) -> Option<i64> {
    let q: &Query = &q.0;
    q.get_timestamp()
}

#[rustler::nif]
fn q_get_tracing(q: ResourceArc<QueryResource>) -> bool {
    let q: &Query = &q.0;
    q.get_tracing()
}

#[rustler::nif]
fn q_new(query_text: String) -> ResourceArc<QueryResource> {
    Query::new(query_text).ex()
}

#[rustler::nif]
fn q_set_consistency(
    q: ResourceArc<QueryResource>,
    consistency: ScyllaConsistency,
) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_consistency(consistency.into());
    q.ex()
}

#[rustler::nif]
fn q_set_is_idempotent(
    q: ResourceArc<QueryResource>,
    is_idempotent: bool,
) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_is_idempotent(is_idempotent);
    q.ex()
}

#[rustler::nif]
fn q_set_page_size(q: ResourceArc<QueryResource>, page_size: i32) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_page_size(page_size);
    q.ex()
}

#[rustler::nif]
fn q_set_retry_policy(_q: ResourceArc<QueryResource>) -> Atom {
    not_implemented_yet()
}

#[rustler::nif]
fn q_set_serial_consistency(
    q: ResourceArc<QueryResource>,
    sc: Option<ScyllaSerialConsistency>,
) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_serial_consistency(sc.map(|sc| sc.into()));
    q.ex()
}

#[rustler::nif]
fn q_set_timestamp(
    q: ResourceArc<QueryResource>,
    timestamp: Option<i64>,
) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_timestamp(timestamp);
    q.ex()
}

#[rustler::nif]
fn q_set_tracing(q: ResourceArc<QueryResource>, should_trace: bool) -> ResourceArc<QueryResource> {
    let mut q: Query = q.0.to_owned();
    q.set_tracing(should_trace);
    q.ex()
}

#[rustler::nif]
fn q_with_page_size(q: ResourceArc<QueryResource>, page_size: i32) -> ResourceArc<QueryResource> {
    let q: Query = q.0.to_owned();
    q.with_page_size(page_size).ex()
}
